copy(): Copy files or folders from a container to a tar stream. copyToDi...

Boran opened the pull request:
After over an hour trying to rebase and solve merge conflicts on #89 I'm giving up on that branch and create this new one 'file-copy2'. All changes are in one commit.

r0b1n commented on src/Docker/Manager/ContainerManager.php:
```diff
@@ -172,6 +172,56 @@ public function start(Container $container, array $hostConfig = [])
+    {
+        $stream = $this->copy($container, $source);
+
+        $output = fopen($destination, 'w+');
```
We need to handle errors like 'permission denied' or 'file not found' (in case directory not exists). I suggest not to use fopen, it throws uncatchable warning in case of opening error.
Quick and dirty solution - mute fopen with `@` and then check $output is not `null`.
Better solution - open file by SPL `SplFileObject` in try-catch block.

ubermuda replied:
+1 for `SplFileObject`

Boran commented:
Apparently Spl is in php 5.x, so just tried:
$fd=SplFileObject("/tmp/foo", 'w+');
and it give "Call to undefined function SplFileObject"
http://php.net/manual/en/spl.configuration.php does not indicate any run time configuration. 
Using Ubuntu 14.04.
Suggestions?

ubermuda commented:
Hey Sean, `SplFileObject` is actually an object, so you need to instantiate it with `new`, like this:

```
$fd = new SplFileObject('/tmp/foo', 'w+');
```

Boran commented:
Oops :-)

Boran commented:
Using fopen rather than SplFileObject  due to stream_copy_to_stream()

Boran commented:
Hi, can this be committed please?

ubermuda commented on src/Docker/Manager/ContainerManager.php:
```diff
@@ -172,6 +172,61 @@ public function start(Container $container, array $hostConfig = [])
+        $stream = $this->copy($container, $source);
+
+        $fd = @fopen($destination, 'w+');
+        if ($fd) {
```
Could you just add an empty line before that `if`?

ubermuda commented on src/Docker/Manager/ContainerManager.php:
```diff
@@ -172,6 +172,61 @@ public function start(Container $container, array $hostConfig = [])
+        if ($fd) {
+          stream_copy_to_stream($stream->detach(), $fd);
+          fclose($fd);
+
```
and remove that one :)

ubermuda commented:
And sorry for being a coding style nazi

ubermuda commented:
Hey @boran, sorry for the delay. Before this can be merged, indentation and CS needs to be fixed:

* use 4 spaces for indentation
* no new line before the closing `}` of an `if`

Boran commented:
I've been waiting 6 months to get this small patch integrated, hardly motivating..

hacfi commented:
The failing test is not related to this PR. Looks good to me :+1:

ubermuda commented:
> I've been waiting 6 months to get this small patch integrated, hardly motivating..

You are right, and I should have handled your PR earlier and in a more swift manner. That being said, there still are inconsistencies in your PR regarding to CS, and you merged master in your branch instead of rebasing your branch on master, which screws up the history.

Anyway, I made a new PR based on your commits, fixed the CS, then rebased against master and squashed the commits in #128

Boran commented:
Oops, apologies for that I had forgotten the correct procedure. Thanks for sorting it out:)

Sean Boran

> On 04 Sep 2015, at 16:32, Geoffrey Bachelet <notifications@github.com> wrote:
> 
> I've been waiting 6 months to get this small patch integrated, hardly motivating..
> 
> You are right, and I should have handled your PR earlier and in a more swift manner. That being said, there still are inconsistencies in your PR regarding to CS, and you merged master in your branch instead of rebasing your branch on master, which screws up the history.
> 
> Anyway, I made a new PR based on your commits, fixed the CS, then rebased against master and squashed the commits in #128
> 
> —
> Reply to this email directly or view it on GitHub.
>

ubermuda commented:
No problem. Just merged #128
